javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -131,11 +134,33 @@ private static int treeSize(int branchCount, int depthShift) {
         } else if (n <= 0) {
             return empty();
         } else {
-            final Object[] root = modifyLeaf(array, depthShift, (offset + n) - 1, Arrays::copyTake, Arrays::copyTake);
-            return new BitMappedTrie<>(root, offset, n, depthShift);
+            final int index = n - 1;
+            final Object[] root = arePointingToSameLeaf(index, length() - 1)
+                                  ? array
+                                  : modifyLeaf(array, depthShift, offset + index, Arrays::copyTake, IDENTITY);
+            return collapsed(root, offset, n, depthShift);
         }
     }
 
+    private boolean arePointingToSameLeaf(int i, int j) {
+        final boolean result = firstDigit(offset + i, BRANCHING_BASE) == firstDigit(offset + j, BRANCHING_BASE);
+        assert result == (getLeaf(i) == getLeaf(j));
+        return result;
+    }
+
+    /* drop root node while it has a single element */
+    private static <T> BitMappedTrie<T> collapsed(Object[] array, int offset, int length, int shift) {
```

Comment:
Since the tree is conceptually not completely symmetric (i.e. a `prepend` changes existing indices, an `append` doesn't), we don't need to store every array as if it were a full array, i.e. if we have only `20` elements, the array can be of length < `32`.
If you have a full array and drop `10` from the end, we can create a new array of size `22`.
If however we drop from the beginning, it's easier to make a full array, and only copy the kept elements (and store the difference in the offset), having the effect of `prepend` with `null`s.
If we would want to eliminate all leading `null`s, the computations would be more difficult (though I can try again, if this constant overhead (that's applicable only to non-leaf nodes) bothers you).